Content/posts/otel-integration.md

```diff
@@ -0,0 +1,303 @@
+    }
+    
+    public func didBootAsync(_ application: Application) async throws {
+        task = Task {
+            try await metrics.run()
+        }
+    }
+    
+    func shutdownAsync(_ application: Application) async {
+        task?.cancel()
+    }
+}
+```
+
+This is a bit of a hack to be able to use the swift-service-lifecycle approach swift-otel uses in Vapor 4, which doesn't implement it.
+
+This is enough to get the queue workers to start exporting metrics to Prometheus via the OpenTelemetry Collector. However we need to configure the collector to actually collect the data and send it to Prometheus. We can do this by creating a `collector-config.yml` file in the same directory as the `docker-compose.yml` file:
+
+```yaml
+# Receive metrics on :4317
+receivers:
+  otlp:
+    protocols:
+      grpc:
+        endpoint: otel-collector:4317
+
+# Expose :7070 for Prometheus to scrape
+exporters:
+  prometheus:
+    endpoint: otel-collector:7070
+    const_labels:
+      collector: "otel-collector"
+
+service:
+  pipelines:
+    metrics:
+      receivers: [otlp]
+      exporters: [prometheus]
+```
+
+And finally add the collector to the `docker-compose.yml` file:
+
+```yaml
+otel-collector:
+  image: otel/opentelemetry-collector-contrib
+  command: ["--config=/etc/otel-collector-config.yml"]
+  ports:
+    - "4317:4317"
+    - "7070:7070"
+  volumes:
+    - "./otel-collector-config.yml:/etc/otel-collector-config.yml"
+```
+
+To be sure that Prometheus gets spun up after the collector, we can add a `depends_on` key to the Prometheus service in the `docker-compose.yml` file:
+
+```yaml
+prometheus:
+  image: prom/prometheus
+  depends_on:
+    - otel-collector
+```
+
+Now, the queue workers should be exporting metrics to the OpenTelemetry Collector, which in turn should be exporting them to Prometheus. Prometheus should be scraping the `/metrics` endpoint on the HTTP server, and the queue workers should be exporting metrics to Prometheus via the OpenTelemetry Collector.
+
+### Visualizing the data
+
+The final step is to visualize the data. For this we'll be using Grafana. Grafana is a tool that allows you to create dashboards to visualize your data. Let's create a simple configuration for Grafana in a `grafana.yml` file:
+
+```yaml
+apiVersion: 1
+
+datasources:
+  - name: Prometheus
+    type: prometheus
+    access: proxy
+    orgId: 1
+    url: http://prometheus:9090
+    isDefault: true
+```
+
+This configuration tells Grafana to use Prometheus as a data source. We can then add the Grafana service to the `docker-compose.yml` file:
+
+```yaml
+grafana:
+  image: grafana/grafana
+  depends_on:
+    - prometheus
+  ports:
+    - "3000:3000" # Grafana UI
+  volumes:
+    - ./grafana.yml:/etc/grafana/provisioning/datasources/grafana.yml
+```
+
+And that's it! Once the Docker Compose stack is running, you should be able to access Grafana at `http://localhost:3000` and start creating dashboards to visualize the data.
```

Comment:
It would be really cool to show an example of some queues metrics here, even if we have to redact queues names